sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala

```diff
@@ -70,8 +70,9 @@ abstract class FileScanBuilder(
   }
 
   override def pushFilters(filters: Seq[Expression]): Seq[Expression] = {
+    val (deterministicFilters, nonDeterminsticFilters) = filters.partition(_.deterministic)
```

Comment:
This logic seems to split too early because `partitionFilters` are not pushed down initially.
Can we split `dataFilters` only?

Reply:
> Are you saying only filter out non-deterministic expressions from being data filters?

Yes. 

>  The idea was to exclude them from both to match the V1 behavior

The idea is that V1 is also not pushing down partition filters technically. We don't need to follow inefficient part from V1, do we?

Reply:
The difference from V1 isn't the only problem. Many built-in non-deterministic functions fail when being pushed down as partition filters (which are evaluated before the execution actually starts I guess) with:

```
java.lang.IllegalArgumentException: requirement failed: Nondeterministic expression org.apache.spark.sql.catalyst.expressions.Rand should be initialized before eval.
```
Same applies to `input_file_name` at least, not sure what else. This is compounded by the fact that expressions referencing no attributes get pushed as "partition" filters too, because an empty set is a subset of every set. I.e. `rand() > 0.5` gets pushed as a partition filter, as does `length(input_file_name()) > 0`, obviously these are very trivial examples but there are more realistic use cases for some things like this.

The no referenced attributes could probably be considered a separate issue/bug, but the problem here still exists. If someone wants to figure out how to make those compatible with partition filters they could, but right now at least some of it is broken so should probably just be blanket filtered out until then?

Reply:
Sorry, but is it relevant to my request, @Kimahriman ?

Reply:
Got it. Could you provide a test case for your example?

Reply:
Got it.